pkg/pmem-csi-driver/pmem-csi-driver.go

```diff
@@ -194,31 +196,40 @@ func (pmemd *pmemDriver) Run() error {
 }
 
 func (pmemd *pmemDriver) registerNodeController() error {
-	fmt.Printf("Connecting to Registry at : %s\n", pmemd.cfg.RegistryEndpoint)
 	var err error
 	var conn *grpc.ClientConn
 
+	req := registry.RegisterControllerRequest{
+		NodeId:   pmemd.cfg.NodeID,
+		Endpoint: pmemd.cfg.ControllerEndpoint,
+	}
+
 	for {
+		glog.Infof("Connecting to registry server at: %s\n", pmemd.cfg.RegistryEndpoint)
 		conn, err = pmemgrpc.Connect(pmemd.cfg.RegistryEndpoint, pmemd.clientTLSConfig, connectionTimeout)
 		if err == nil {
-			glog.Info("Connected to RegistryServer!!!")
 			break
 		}
-		/* TODO: Retry loop */
-		glog.Infof("Failed to connect RegistryServer : %s, retrying after 10 seconds...", err.Error())
-		time.Sleep(10 * time.Second)
+		glog.Infof("Failed to connect registry server: %s, retrying after %v seconds...", err.Error(), retryTimeout.Seconds())
+		time.Sleep(retryTimeout)
 	}
-	client := registry.NewRegistryClient(conn)
-	req := registry.RegisterControllerRequest{
-		NodeId:   pmemd.cfg.NodeID,
-		Endpoint: pmemd.cfg.ControllerEndpoint,
-	}
-	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
-	defer cancel()
-
-	if _, err = client.RegisterController(ctx, &req); err != nil {
-		return fmt.Errorf("Fail to register with server at '%s' : %s", pmemd.cfg.RegistryEndpoint, err.Error())
+	for {
+		client := registry.NewRegistryClient(conn)
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
+		defer cancel()
```

Comment:
`cancel` is only called when when the function returns. That means each loop iteration ends up allocating more resources. You need to call `cancel()` after `client.RegisterController`, regardless whether the call succeeded or failed.

Reply:
Yes, using `defer` inside loop is bad idea. I think we can start loop after calling `defer cancel() `as we can reuse `client` and `ctx`.